cmd/soroban-rpc/internal/preflight/pool.go

```diff
@@ -0,0 +1,72 @@
+package preflight
+
+import (
+	"context"
+	"sync"
+
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-tools/cmd/soroban-rpc/internal/db"
+)
+
+type workerResult struct {
+	preflight Preflight
+	err       error
+}
+
+type workerRequest struct {
+	ctx        context.Context
+	params     PreflightParameters
+	resultChan chan<- workerResult
+}
+
+type PreflightWorkerPool struct {
+	ledgerEntryReader db.LedgerEntryReader
+	networkPassphrase string
+	logger            *log.Entry
+	requestChan       chan workerRequest
+	wg                sync.WaitGroup
+}
+
+func NewPreflightWorkerPool(workerCount uint, ledgerEntryReader db.LedgerEntryReader, networkPassphrase string, logger *log.Entry) *PreflightWorkerPool {
+	result := PreflightWorkerPool{
+		ledgerEntryReader: ledgerEntryReader,
+		networkPassphrase: networkPassphrase,
+		logger:            logger,
+		requestChan:       make(chan workerRequest),
+	}
+	for i := uint(0); i < workerCount; i++ {
+		result.wg.Add(1)
+		go result.work()
+	}
+	return &result
+}
+
+func (pwp *PreflightWorkerPool) work() {
+	defer pwp.wg.Done()
+	for request := range pwp.requestChan {
+		preflight, err := GetPreflight(request.ctx, request.params)
+		request.resultChan <- workerResult{preflight, err}
+	}
+}
+
+func (pwp *PreflightWorkerPool) Close() {
+	close(pwp.requestChan)
+	pwp.wg.Wait()
+}
+
+func (pwp *PreflightWorkerPool) GetPreflight(ctx context.Context, sourceAccount xdr.AccountId, op xdr.InvokeHostFunctionOp) (Preflight, error) {
+	params := PreflightParameters{
+		Logger:             pwp.logger,
+		SourceAccount:      sourceAccount,
+		InvokeHostFunction: op,
+		NetworkPassphrase:  pwp.networkPassphrase,
+		LedgerEntryReader:  pwp.ledgerEntryReader,
+	}
+	resultC := make(chan workerResult)
+	defer close(resultC)
+	pwp.requestChan <- workerRequest{ctx, params, resultC}
+	result := <-resultC
```

Comment:
due to the above issue, this line might be hunged up forever..

Reply:
closed channels return inmediately

Reply:
I don't think that resultC can be a closed channel here. ( nor does the `defer close(resultC)` really needed ).
But you're correct that in the current code, it can't hang:
- if the write to pwp.requestChan was successful, it means that the channel is still open, and there is a processing goroutine.
- if there is a processing go routine, then resultC should get called, eventually.

Reply:
When saying _closed channels return immediately_ I was referring to `pwp.requestChan`

And, you are right, I forgot unused (and unreachable) channels are garbage-collected. I removed the close.